azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+        # Given a priority queue of files to collect, add one by one while the archive size is under the size limit.
+        # If a single file is over the file size limit, truncate it before adding it to the archive.
+        self._log_to_results_file("\n### Preparing list of files to add to archive ###")
+        total_uncompressed_size = 0
+        final_files_to_collect = []
+
+        while priority_file_queue:
+            file_path = heappop(priority_file_queue)[1]  # (priority, file_path)
+            file_size = min(os.path.getsize(file_path), FILE_SIZE_LIMIT)
+
+            if total_uncompressed_size + file_size > ARCHIVE_SIZE_LIMIT:
+                self._log_to_results_file("Archive too big, done with adding files.")
+                break
+
+            if os.path.getsize(file_path) <= FILE_SIZE_LIMIT:
+                final_files_to_collect.append(file_path)
+                self._log_to_results_file("Adding file {0}, size {1}b".format(file_path, file_size))
+            else:
+                truncated_file_path = self._truncate_large_file(file_path)
+                if truncated_file_path:
+                    self._log_to_results_file("Adding truncated file {0}, size {1}b".format(truncated_file_path, file_size))
+                    final_files_to_collect.append(truncated_file_path)
+
+            total_uncompressed_size += file_size
+
+        return final_files_to_collect
+
+    def _create_list_of_files_to_collect(self):
+        # The final list of files to be collected by zip is created in three steps:
+        # 1) Parse given manifest file, expanding wildcards and keeping a list of files that exist on disk
+        # 2) Assign those files a priority depending on whether they are in the must collect file list.
+        # 3) In priority order, add files to the final list to be collected, until the size of the archive is under
+        #    the size limit.
+        parsed_file_paths = self._parse_manifest_file()
+        prioritized_file_paths = self._get_priority_files_list(parsed_file_paths)
+        files_to_collect = self._get_final_list_for_archive(prioritized_file_paths)
+        return files_to_collect
+
+    def collect_logs(self):
+        """
+        Public method that collects necessary log files in a tarball that is updated each time this method is invoked.
+        The tarball is then compressed into a zip.
+        :return: Returns True if the log collection succeeded
+        """
+        try:
+            # Clear previous run's output and create base directories if they don't exist already
+            rm_files(OUTPUT_RESULTS_FILE_PATH)
+            mkdir(TRUNCATED_FILES_DIR)
+            mkdir(LOG_COLLECTOR_DIR)
+
+            files_to_collect = self._create_list_of_files_to_collect()
+            self._log_to_results_file("\n### Creating archive ###")
+
+            self._remove_deleted_files_from_archive(files_to_collect)
+            self._update_files_in_archive(files_to_collect)
```

Comment:
After we decided which files to collect, the archive is created. Since we don't recreate the entire archive every single collection call (as there will be many of the same files to collect), we only remove the files that have been deleted from disk since the last collection call, add new files and update existing files if they've been modified.

Reply:
We expect the difference in files between two collection times (given a collection interval of 1h) to be mostly new files (new GoalState.xml, etc.) or updated files (any log file will have newer entries).

Most of the files will remain the same (for example, old GoalState.xml of Config.xml files), so we avoid collecting them again every single call. They are kept in the archive from last time.